lib/mwUtil.js

```diff
@@ -311,7 +311,8 @@ mwUtil.createRelativeTitleRedirect = (path, req, options) => {
             + new URI(pathPatternAfterTitle, newReqParams, true).toString().substr(1)
             + mwUtil.getQueryString(req);
     } else {
-        location = backString + encodeURIComponent(newReqParams.title) + mwUtil.getQueryString(req);
+        location = backString + encodeURIComponent(newReqParams.title).replace(/%23/g, '#')
+            + mwUtil.getQueryString(req);
```

Comment:
I believe the order is incorrect - everything after `#` is client-side, so the query string must be before the hashbang